File: monitor.py
#!/usr/bin/env python3
"""
URL change monitor.
Fetches each URL, hashes its content, and compares against the last known hash.
Sends a Slack notification when a change is detected.
Hashes are stored as files in .cache/ so GitHub Actions can persist them via cache action.

Some URLs use custom extractors instead of full-page hashing — these extract a
specific signal from the page (e.g. only security-tagged releases) so that
unrelated page changes don't trigger false positives.
"""

import hashlib
import json
import os
import re
import sys
import urllib.request
import urllib.error
from datetime import datetime, timezone
from pathlib import Path

CACHE_DIR = Path(".cache")
SLACK_WEBHOOK_URL = os.environ.get("SLACK_WEBHOOK_URL", "")

# Set to True to print Slack messages without actually sending them.
SLACK_DRY_RUN = True


def fetch_content(url: str) -> str | None:
    """Fetch the text content of a URL. Returns None on failure."""
    try:
        req = urllib.request.Request(url, headers={"User-Agent": "Mozilla/5.0 (URL Monitor)"})
        with urllib.request.urlopen(req, timeout=15) as response:
            return response.read().decode("utf-8", errors="replace")
    except urllib.error.URLError as e:
        print(f"  [ERROR] Could not fetch {url}: {e}")
        return None


def hash_content(content: str) -> str:
    return hashlib.sha256(content.encode("utf-8")).hexdigest()


CACHE_NAMES: dict[str, str] = {
    "https://www.shopware.com/nl/changelog/": "shopware-6-changelog.json",
    "https://store.shopware.com/en/swag136939272659f/shopware-6-security-plugin.html": "shopware-6-security-plugin-changelog.json",
}


def cache_path(url: str) -> Path:
    """Return a stable file path for caching the hash of a URL."""
    name = CACHE_NAMES.get(url) or f"{hashlib.md5(url.encode()).hexdigest()}.json"
    return CACHE_DIR / name


def load_cached(url: str) -> dict | None:
    path = cache_path(url)
    if path.exists():
        return json.loads(path.read_text())
    return None


def save_cache(url: str, data: dict):
    CACHE_DIR.mkdir(exist_ok=True)
    path = cache_path(url)
    path.write_text(json.dumps({
        "url": url,
        "last_checked": datetime.now(timezone.utc).isoformat(),
        **data,
    }, indent=2))


def _now() -> str:
    return datetime.now(timezone.utc).strftime("%Y-%m-%d %H:%M UTC")


def send_slack_notification(url: str, message: str):
    if SLACK_DRY_RUN:
        print(f"  [DRY RUN] Would send to Slack:\n{message}")
        return

    if not SLACK_WEBHOOK_URL:
        print("  [WARN] SLACK_WEBHOOK_URL not set — skipping notification.")
        return

    payload = json.dumps({"text": message}).encode("utf-8")
    req = urllib.request.Request(
        SLACK_WEBHOOK_URL,
        data=payload,
        headers={"Content-Type": "application/json"},
        method="POST",
    )
    try:
        with urllib.request.urlopen(req, timeout=10) as resp:
            print(f"  [SLACK] Notification sent (status {resp.status}).")
    except urllib.error.URLError as e:
        print(f"  [ERROR] Failed to send Slack notification: {e}")


# ---------------------------------------------------------------------------
# Custom extractors
# ---------------------------------------------------------------------------

def extract_shopware_security_versions(content: str) -> dict[str, str]:
    """
    Return the latest security-tagged release per major Shopware branch.

    The changelog page lists releases newest-first. For each version entry that
    carries the 'release-title--security' badge, we record it as the latest
    security release for its major branch (e.g. "6.5", "6.6", "6.7") — taking
    the first (newest) match per branch.

    Returns a dict like {"6.5": "6.5.8.18", "6.6": "6.6.10.15", "6.7": "6.7.8.1"}.
    """
    version_positions = [
        (m.start(), m.group(1))
        for m in re.finditer(r'release-title--version[^>]*>(\d+\.\d+\.\d+(?:\.\d+)?)', content)
    ]
    latest_per_branch: dict[str, str] = {}
    for i, (pos, version) in enumerate(version_positions):
        next_pos = version_positions[i + 1][0] if i + 1 < len(version_positions) else pos + 5000
        segment = content[pos:min(pos + 3000, next_pos)]
        if "release-title--security" in segment:
            branch = ".".join(version.split(".")[:2])  # "6.5", "6.6", "6.7"
            if branch not in latest_per_branch:
                latest_per_branch[branch] = version
    return dict(sorted(latest_per_branch.items()))


def extract_shopware_security_plugin_versions(content: str) -> dict[str, str]:
    """
    Return the latest plugin version per Shopware branch from the store changelog.

    All entries are security releases. The plugin major version maps to a branch:
      2.x → 6.5,  3.x → 6.6,  4.x → 6.7

    Returns a dict like {"6.5": "2.0.19", "6.6": "3.0.14", "6.7": "4.0.9"}.
    """
    branch_map = {"2": "6.5", "3": "6.6", "4": "6.7"}
    latest_per_branch: dict[str, str] = {}
    for m in re.finditer(r'<h3[^>]*class="changelogs-header"[^>]*>\s*(\d+\.\d+\.\d+)\s', content, re.DOTALL):
        version = m.group(1)
        major = version.split(".")[0]
        branch = branch_map.get(major)
        if branch and branch not in latest_per_branch:
            latest_per_branch[branch] = version
    return dict(sorted(latest_per_branch.items()))


# Maps a URL to a custom extractor function. Extractors receive the raw HTML
# and return a JSON-serialisable value stored in the cache and compared across
# runs. Only the extracted signal triggers notifications.
EXTRACTORS: dict[str, callable] = {
    "https://www.shopware.com/nl/changelog/": extract_shopware_security_versions,
    "https://store.shopware.com/en/swag136939272659f/shopware-6-security-plugin.html": extract_shopware_security_plugin_versions,
}


# ---------------------------------------------------------------------------
# Check logic
# ---------------------------------------------------------------------------

def check_url(url: str):
    print(f"\nChecking: {url}")
    content = fetch_content(url)
    if content is None:
        return

    extractor = EXTRACTORS.get(url)
    if extractor:
        _check_url_with_extractor(url, content, extractor)
    else:
        _check_url_hash(url, content)


def _check_url_hash(url: str, content: str):
    current_hash = hash_content(content)
    cached = load_cached(url)

    if cached is None:
        print("  [NEW] No previous record found — saving baseline.")
        save_cache(url, {"hash": current_hash})
        send_slack_notification(url, f":eyes: *URL Monitor* is now tracking:\n<{url}|{url}>\n_First check completed at {_now()} — future changes will trigger a notification._")

    elif cached.get("hash") != current_hash:
        print(f"  [CHANGED] Hash mismatch detected!")
        print(f"    Old: {cached['hash'][:16]}…")
        print(f"    New: {current_hash[:16]}…")
        save_cache(url, {"hash": current_hash})
        send_slack_notification(url, f":bell: *Page changed!*\n<{url}|{url}>\n_Detected at {_now()}_")

    else:
        print(f"  [OK] No change detected (hash: {current_hash[:16]}…)")


def _check_url_with_extractor(url: str, content: str, extractor):
    current_value = extractor(content)
    cached = load_cached(url)

    if cached is None or "extracted" not in cached:
        print(f"  [NEW] No previous record found — saving baseline.")
        print(f"    Extracted: {current_value}")
        save_cache(url, {"extracted": current_value})
        send_slack_notification(url, f":eyes: *URL Monitor* is now tracking:\n<{url}|{url}>\n_Baseline set at {_now()}._")
        return

    previous_value = cached["extracted"]

    if isinstance(current_value, dict) and isinstance(previous_value, dict):
        changed = {
            branch: (previous_value.get(branch), new_ver)
            for branch, new_ver in current_value.items()
            if new_ver != previous_value.get(branch)
        }
        if changed:
            print(f"  [CHANGED] New security versions: {changed}")
            save_cache(url, {"extracted": current_value})
            lines = "\n".join(
                f"  • {branch}: `{new}`  (was `{old}`)"
                for branch, (old, new) in sorted(changed.items())
            )
            send_slack_notification(url, f":rotating_light: *Shopware Security Update(s) released!*\n<{url}|{url}>\n{lines}\n_Detected at {_now()}_")
        else:
            print(f"  [OK] No new security versions: {current_value}")
    else:
        if current_value != previous_value:
            print(f"  [CHANGED] Extracted value changed.")
            save_cache(url, {"extracted": current_value})
            send_slack_notification(url, f":bell: *Page changed!*\n<{url}|{url}>\n_Detected at {_now()}_")
        else:
            print(f"  [OK] No change detected.")


if __name__ == "__main__":
    urls = sys.argv[1:]
    if not urls:
        print("Usage: python3 monitor.py <url1> [url2] ...")
        sys.exit(1)

    for url in urls:
        check_url(url)

    print("\nDone.")

File: monitor.php
<?php

declare(strict_types=1);

/**
 * URL change monitor.
 *
 * Fetches each URL, hashes its content, and compares against the last known hash.
 * Sends a Slack notification when a change is detected.
 * Hashes are stored as JSON files in .cache/ so GitHub Actions can persist them
 * via the cache commit step.
 *
 * Some URLs use custom extractors instead of full-page hashing. These extract a
 * specific signal from the page (e.g. only security-tagged releases) so that
 * unrelated page changes don't trigger false positives.
 */
class Monitor
{
    public const CACHE_DIR = '.cache';

    // Set to true to print Slack messages without actually sending them.
    public const SLACK_DRY_RUN = true;

    // Maps URLs to the name of their custom extractor method on this class.
    private const EXTRACTORS = [
        'https://www.shopware.com/nl/changelog/' => 'extractShopwareSecurityVersions',
    ];

    // -------------------------------------------------------------------------
    // Entry point
    // -------------------------------------------------------------------------

    public static function run(array $urls): void
    {
        foreach ($urls as $url) {
            self::checkUrl($url);
        }
        echo "\nDone.\n";
    }

    // -------------------------------------------------------------------------
    // Check logic
    // -------------------------------------------------------------------------

    public static function checkUrl(string $url): void
    {
        echo "\nChecking: {$url}\n";
        $content = self::fetchContent($url);
        if ($content === null) {
            return;
        }

        $extractor = self::EXTRACTORS[$url] ?? null;
        if ($extractor !== null) {
            self::checkUrlWithExtractor($url, $content, $extractor);
        } else {
            self::checkUrlHash($url, $content);
        }
    }

    private static function checkUrlHash(string $url, string $content): void
    {
        $currentHash = self::hashContent($content);
        $cached      = self::loadCached($url);

        if ($cached === null) {
            echo "  [NEW] No previous record found — saving baseline.\n";
            self::saveCache($url, ['hash' => $currentHash]);
            self::sendSlackNotification(
                $url,
                ":eyes: *URL Monitor* is now tracking:\n<{$url}|{$url}>\n"
                . '_First check completed at ' . self::now() . ' — future changes will trigger a notification._'
            );
        } elseif (($cached['hash'] ?? null) !== $currentHash) {
            echo "  [CHANGED] Hash mismatch detected!\n";
            echo '    Old: ' . substr($cached['hash'] ?? '', 0, 16) . "…\n";
            echo '    New: ' . substr($currentHash, 0, 16) . "…\n";
            self::saveCache($url, ['hash' => $currentHash]);
            self::sendSlackNotification(
                $url,
                ":bell: *Page changed!*\n<{$url}|{$url}>\n_Detected at " . self::now() . '_'
            );
        } else {
            echo '  [OK] No change detected (hash: ' . substr($currentHash, 0, 16) . "…)\n";
        }
    }

    private static function checkUrlWithExtractor(string $url, string $content, string $extractorMethod): void
    {
        $currentValue = call_user_func([self::class, $extractorMethod], $content);
        $cached       = self::loadCached($url);

        if ($cached === null || !array_key_exists('extracted', $cached)) {
            echo "  [NEW] No previous record found — saving baseline.\n";
            echo '    Extracted: ' . json_encode($currentValue) . "\n";
            self::saveCache($url, ['extracted' => $currentValue]);
            self::sendSlackNotification(
                $url,
                ":eyes: *URL Monitor* is now tracking:\n<{$url}|{$url}>\n_Baseline set at " . self::now() . '._'
            );
            return;
        }

        $previousValue = $cached['extracted'];

        if (is_array($currentValue) && is_array($previousValue)) {
            $changed = [];
            foreach ($currentValue as $branch => $newVer) {
                $oldVer = $previousValue[$branch] ?? null;
                if ($newVer !== $oldVer) {
                    $changed[$branch] = [$oldVer, $newVer];
                }
            }

            if (!empty($changed)) {
                echo '  [CHANGED] New security versions: ' . json_encode($changed) . "\n";
                self::saveCache($url, ['extracted' => $currentValue]);
                ksort($changed);
                $lines = implode("\n", array_map(
                    fn ($branch, $pair) => "  • {$branch}: `{$pair[1]}`  (was `{$pair[0]}`)",
                    array_keys($changed),
                    array_values($changed)
                ));
                self::sendSlackNotification(
                    $url,
                    ":rotating_light: *Shopware Security Update(s) released!*\n<{$url}|{$url}>\n"
                    . $lines . "\n_Detected at " . self::now() . '_'
                );
            } else {
                echo '  [OK] No new security versions: ' . json_encode($currentValue) . "\n";
            }
        } else {
            if ($currentValue !== $previousValue) {
                echo "  [CHANGED] Extracted value changed.\n";
                self::saveCache($url, ['extracted' => $currentValue]);
                self::sendSlackNotification(
                    $url,
                    ":bell: *Page changed!*\n<{$url}|{$url}>\n_Detected at " . self::now() . '_'
                );
            } else {
                echo "  [OK] No change detected.\n";
            }
        }
    }

    // -------------------------------------------------------------------------
    // Custom extractors
    // -------------------------------------------------------------------------

    /**
     * Return the latest security-tagged release per major Shopware branch.
     *
     * The changelog page lists releases newest-first, so the first security
     * entry found per major branch (e.g. "6.5", "6.6", "6.7") is the latest.
     *
     * Returns e.g. ['6.5' => '6.5.8.18', '6.6' => '6.6.10.15', '6.7' => '6.7.8.1'].
     */
    public static function extractShopwareSecurityVersions(string $content): array
    {
        preg_match_all(
            '/release-title--version[^>]*>(\d+\.\d+\.\d+(?:\.\d+)?)/',
            $content,
            $allMatches,
            PREG_SET_ORDER | PREG_OFFSET_CAPTURE
        );

        $versionPositions = [];
        foreach ($allMatches as $match) {
            $versionPositions[] = [
                'pos'     => $match[0][1],  // byte offset of full match
                'version' => $match[1][0],  // captured version string
            ];
        }

        $latestPerBranch = [];
        foreach ($versionPositions as $i => $item) {
            $pos     = $item['pos'];
            $version = $item['version'];
            $nextPos = $versionPositions[$i + 1]['pos'] ?? $pos + 5000;
            $segment = substr($content, $pos, min(3000, $nextPos - $pos));

            if (str_contains($segment, 'release-title--security')) {
                $parts  = explode('.', $version);
                $branch = $parts[0] . '.' . $parts[1];  // e.g. "6.7"
                if (!isset($latestPerBranch[$branch])) {
                    $latestPerBranch[$branch] = $version;
                }
            }
        }

        ksort($latestPerBranch);

        return $latestPerBranch;
    }

    // -------------------------------------------------------------------------
    // HTTP
    // -------------------------------------------------------------------------

    public static function fetchContent(string $url): ?string
    {
        $context = stream_context_create([
            'http' => [
                'timeout'        => 15,
                'header'         => 'User-Agent: Mozilla/5.0 (URL Monitor)',
                'ignore_errors'  => true,
            ],
        ]);

        $content = @file_get_contents($url, false, $context);

        if ($content === false) {
            $error = error_get_last()['message'] ?? 'unknown error';
            echo "  [ERROR] Could not fetch {$url}: {$error}\n";
            return null;
        }

        return $content;
    }

    // -------------------------------------------------------------------------
    // Cache
    // -------------------------------------------------------------------------

    public static function hashContent(string $content): string
    {
        return hash('sha256', $content);
    }

    public static function cachePath(string $url): string
    {
        return self::CACHE_DIR . '/' . hash('md5', $url) . '.json';
    }

    public static function loadCached(string $url): ?array
    {
        $path = self::cachePath($url);
        if (!file_exists($path)) {
            return null;
        }
        return json_decode(file_get_contents($path), true);
    }

    public static function saveCache(string $url, array $data): void
    {
        if (!is_dir(self::CACHE_DIR)) {
            mkdir(self::CACHE_DIR, 0755, true);
        }
        $payload = array_merge(
            [
                'url'          => $url,
                'last_checked' => (new DateTime('now', new DateTimeZone('UTC')))->format(DateTime::ATOM),
            ],
            $data
        );
        file_put_contents(
            self::cachePath($url),
            json_encode($payload, JSON_PRETTY_PRINT | JSON_UNESCAPED_SLASHES)
        );
    }

    // -------------------------------------------------------------------------
    // Notifications
    // -------------------------------------------------------------------------

    public static function sendSlackNotification(string $url, string $message): void
    {
        if (self::SLACK_DRY_RUN) {
            echo "  [DRY RUN] Would send to Slack:\n{$message}\n";
            return;
        }

        $webhookUrl = getenv('SLACK_WEBHOOK_URL') ?: '';
        if ($webhookUrl === '') {
            echo "  [WARN] SLACK_WEBHOOK_URL not set — skipping notification.\n";
            return;
        }

        $payload = json_encode(['text' => $message]);
        $context = stream_context_create([
            'http' => [
                'method'        => 'POST',
                'timeout'       => 10,
                'header'        => "Content-Type: application/json\r\nContent-Length: " . strlen($payload),
                'content'       => $payload,
                'ignore_errors' => true,
            ],
        ]);

        $response = @file_get_contents($webhookUrl, false, $context);
        if ($response === false) {
            $error = error_get_last()['message'] ?? 'unknown error';
            echo "  [ERROR] Failed to send Slack notification: {$error}\n";
        } else {
            echo "  [SLACK] Notification sent.\n";
        }
    }

    // -------------------------------------------------------------------------
    // Helpers
    // -------------------------------------------------------------------------

    private static function now(): string
    {
        return (new DateTime('now', new DateTimeZone('UTC')))->format('Y-m-d H:i') . ' UTC';
    }
}

// Run when invoked directly from the CLI; skipped when included by tests.
if (!defined('RUNNING_TESTS')) {
    $urls = array_slice($argv, 1);
    if (empty($urls)) {
        echo "Usage: php monitor.php <url1> [url2] ...\n";
        exit(1);
    }
    Monitor::run($urls);
}
